sdk/cxx/include/sawtooth_sdk.h

```diff
@@ -0,0 +1,258 @@
+    //      addresses.push(std::make_pair(address, value));
+    virtual void SetState(const std::vector<KeyValue>& addresses) const = 0;
+
+    // Delete a single GlobalState value. Used the batch Delete function
+    // defined below if you are deleting multiple items.
+    virtual void DeleteState(const std::string& address) const = 0;
+
+    // Delete multiple entries from global state.
+    virtual void DeleteState(const std::vector<std::string>& address) const = 0;
+};
+typedef std::shared_ptr<GlobalStateIF> GlobalStatePtr;
+typedef std::unique_ptr<GlobalStateIF> GlobalStateUPtr;
+
+
+// TransactionApplicator is an interface that defined to represent the
+// processing of a single transaction. This object should make no assumptions
+// about the thread and timing of when it will execute. The
+// TransactionApplicator owns the reference to the transaction and the global
+// state it is passed.
+class TransactionApplicator {
+ public:
+    TransactionApplicator(sawtooth::TransactionUPtr txn,
+        sawtooth::GlobalStateUPtr state) : txn(std::move(txn)),
+        state(std::move(state)) {}
+    virtual ~TransactionApplicator() {}
+
+
+    // Process the transaction, this function should do all of the processing
+    // of the given transaction. The general framework should be:
+    // 1) Validate the transaction
+    // 2) retrieve the state from GlobalState
+    // 3) validate the transaction changes are valid
+    // 4) Write the updated state to GlobalState
+    //
+    // Any exception thrown from this function will result in the transaction
+    // being marked as invalid.
+    virtual void Apply() = 0;
+ protected:
+    sawtooth::TransactionUPtr txn;
+    sawtooth::GlobalStateUPtr state;
+};
+typedef std::unique_ptr<TransactionApplicator> TransactionApplicatorUPtr;
+
+
+
+// Definition of a TransactionHandler to be registered with the validator.
+// Every Transaction Processor should register at least one of these. This
+// definition allows for the handler to declare multiple version.
+// that it supports.
+class TransactionHandler {
+ public:
+    virtual ~TransactionHandler() {};
+
+    virtual std::string transaction_family_name() const = 0;
+    virtual std::list<std::string> versions() const = 0;
+    virtual std::list<std::string> namespaces() const = 0;
+    // When a Transaction processing reqquest is received, the
+    // TransactionHandler is called with the Transaction and the GlobalState
+    // context to be processed. The Handler should return a subclassed instance
+    // of TransactionApplicator that will handle the processing. As little work
+    // as possible should be done in this function.
+    virtual TransactionApplicatorUPtr GetApplicator(
+        TransactionUPtr txn,
+        GlobalStateUPtr state) = 0;
+};
+typedef std::unique_ptr<TransactionHandler> TransactionHandlerUPtr;
+typedef std::shared_ptr<TransactionHandler> TransactionHandlerPtr;
+
+
+// Helper class to provide mappings from domain identifiers
```

Comment:
AddressMapper and AddressMapperIF are inconsistent with the other SDKs.

For the other SDKs, we are currently designing the specific approach to suggest to users. It will likely not have a generic key/value interface as presented here (which is only interesting to very simplistic mappings) and instead present an approach for mapping the state object with a TF-specific API.  Suggest renaming AddressMapper to IntkeyAddressMapper and moving it to the intkey example.

Reply:
Unclear which interfaces in the other sdks you are refering to. It rather hard to guess what design work is in progress. This has been present in the CXX  SDK since the first revision. This feed back would have been better suited to the original review, rather then on a small incremental improvement.